src/Razor/Microsoft.AspNetCore.Razor.Language/src/Legacy/SeekableTextReader.cs

```diff
@@ -3,32 +3,36 @@
 
 using System;
 using System.IO;
+using System.Text;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
 
 namespace Microsoft.AspNetCore.Razor.Language.Legacy
 {
     internal class SeekableTextReader : TextReader, ITextDocument
     {
-        private readonly LineTrackingStringBuffer _buffer;
+        private readonly RazorSourceDocument _sourceDocument;
         private int _position = 0;
         private int _current;
         private SourceLocation _location;
+        private (TextSpan Span, int LineIndex) _cachedLineInfo;
 
-        public SeekableTextReader(string source, string filePath) : this(source.ToCharArray(), filePath) { }
+        public SeekableTextReader(string source, string filePath) : this(new StringSourceDocument(source, Encoding.UTF8, new RazorSourceDocumentProperties(filePath, relativePath: null))) { }
 
-        public SeekableTextReader(char[] source, string filePath)
+        public SeekableTextReader(RazorSourceDocument source)
         {
             if (source == null)
             {
                 throw new ArgumentNullException(nameof(source));
             }
 
-            _buffer = new LineTrackingStringBuffer(source, filePath);
+            _sourceDocument = source;
+            _cachedLineInfo = (new TextSpan(0, _sourceDocument.Lines.GetLineLength(0)), 0);
```

Comment:
This is confusing to me. Why cache the very first line and special case that?

Reply:
It simplified the code so it didnt need to check against uninitialized. Also, callers tend to start at the first line.

Reply:
This assumes that RazorSourceDocument will have atleast one line right? Is it safe to make that assumption?

Reply:
On empty documents, _sourceDocument.Lines.Count == 1. See [https://github.com/dotnet/aspnetcore/blob/66dd4919dafcb21c57a923a661cdbb0ff351713d/src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorSourceLineCollection.cs#L71](https://github.com/dotnet/aspnetcore/blob/66dd4919dafcb21c57a923a661cdbb0ff351713d/src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorSourceLineCollection.cs#L71)


---
In reply to: [440388134](https://github.com/dotnet/aspnetcore/pull/22903#discussion_r440388134) [](ancestors = 440388134)